File: javascript/taskManager.js
// Import required modules
const fs = require('fs');
const readline = require('readline');

const fileName = 'tasks.json';
let id = 1;

// Initialize ID based on existing tasks
function initializeId() {
    if (fs.existsSync(fileName)) {
        const data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        const tasks = data.tasks || [];
        if (tasks.length > 0) {
            id = Math.max(...tasks.map(task => task.ID)) + 1;
            console.log(`Initialized ID to ${id} based on existing tasks.`);
        } else {
            console.log('No tasks found in JSON file. Using default ID = 1.');
        }
    } else {
        console.log('JSON file not found. Starting fresh with ID = 1.');
    }
}

// Add a new task
function addTask(description, mark = 'todo') {
    const newTask = { description, ID: id++, mark };

    let data = { tasks: [] };
    if (fs.existsSync(fileName)) {
        data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
    }
    data.tasks.push(newTask);
    fs.writeFileSync(fileName, JSON.stringify(data, null, 4));
    console.log('New task added:', newTask);
}

// Update a task
function updateTask(taskId, newDescription) {
    if (fs.existsSync(fileName)) {
        const data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        const task = data.tasks.find(task => task.ID === taskId);

        if (task) {
            task.description = newDescription;
            fs.writeFileSync(fileName, JSON.stringify(data, null, 4));
            console.log(`Task ID ${taskId} updated successfully.`);
        } else {
            console.log(`Task ID ${taskId} not found.`);
        }
    } else {
        console.log('Error: Task file not found. Please add a task first.');
    }
}

// Delete a task
function deleteTask(taskId) {
    if (fs.existsSync(fileName)) {
        const data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        const originalLength = data.tasks.length;
        data.tasks = data.tasks.filter(task => task.ID !== taskId);

        if (data.tasks.length < originalLength) {
            fs.writeFileSync(fileName, JSON.stringify(data, null, 4));
            console.log(`Task ID ${taskId} deleted successfully.`);
        } else {
            console.log(`Task ID ${taskId} not found.`);
        }
    } else {
        console.log('Error: Task file not found. Please create a task first.');
    }
}

// Mark a task
function markTask(taskId, mark) {
    if (fs.existsSync(fileName)) {
        const data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        const task = data.tasks.find(task => task.ID === taskId);

        if (task) {
            task.mark = mark;
            fs.writeFileSync(fileName, JSON.stringify(data, null, 4));
            console.log(`Task ID ${taskId} marked as '${mark}' successfully.`);
        } else {
            console.log(`Task ID ${taskId} not found.`);
        }
    } else {
        console.log('Error: Task file not found. Please add a task first.');
    }
}

// Show tasks
function showTasks(filter = 'all') {
    if (fs.existsSync(fileName)) {
        const data = JSON.parse(fs.readFileSync(fileName, 'utf8'));
        const tasks = data.tasks;

        let filteredTasks = tasks;
        if (filter !== 'all') {
            filteredTasks = tasks.filter(task => task.mark === filter);
        }

        filteredTasks.forEach((task, index) => {
            console.log(`${index + 1}.`, task);
        });
    } else {
        console.log('Error: Task file not found. Please add a task first.');
    }
}


function initializeTasksFile() {
    const data = { tasks: [] }; // Define the structure of the tasks file
    fs.writeFileSync(fileName, JSON.stringify(data, null, 4)); // Write the data to the file with indentation
    console.log('Initialized tasks.json with an empty tasks array.');
}

// CLI for user interaction
function taskCLI() {
    initializeId();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    rl.setPrompt('task-cli> ');
    rl.prompt();

    rl.on('line', input => {
        const args = input.trim().split(/\s+/);
        const command = args[0];

        switch (command) {
            case 'add':
                addTask(args.slice(1).join(' '));
                break;
            case 'update':
                updateTask(parseInt(args[1]), args.slice(2).join(' '));
                break;
            case 'delete':
                deleteTask(parseInt(args[1]));
                break;
            case 'mark-in-progress':
                markTask(parseInt(args[1]), 'in-progress');
                break;
            case 'mark-done':
                markTask(parseInt(args[1]), 'done');
                break;
            case 'list':
                showTasks(args[1] || 'all');
                break;
            case 'exit':
                console.log('Exiting task CLI.');
                rl.close();
                return;
            default:
                console.log('Unknown command. Available commands: add, update, delete, mark-in-progress, mark-done, list, exit.');
        }

        rl.prompt();
    });
}

// Run the CLI if this file is executed directly
if (require.main === module) {
    taskCLI();
}

module.exports = { initializeId, addTask, updateTask, deleteTask, markTask, showTasks, initializeTasksFile};

File: TaskCLI/src/TaskCLI/TaskManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using Newtonsoft.Json;

namespace TaskCLI
{
    public class Task
    {
        public int ID { get; set; }
        public string Description { get; set; }
        public string Mark { get; set; }
    }

    public class TaskManager
    {
        private static readonly string FileName = "tasks.json";
        private static int _id = 1;

        public TaskManager()
        {
            InitializeID();
        }

        private void InitializeID()
        {
            if (File.Exists(FileName))
            {
                try
                {
                    var data = JsonConvert.DeserializeObject<Dictionary<string, List<Task>>>(File.ReadAllText(FileName));
                    if (data != null && data.ContainsKey("tasks") && data["tasks"].Count > 0)
                    {
                        _id = data["tasks"].Max(task => task.ID) + 1;
                        Console.WriteLine($"Initialized ID to {_id}.");
                    }
                }
                catch
                {
                    Console.WriteLine("Invalid JSON file. Starting fresh.");
                }
            }
            else
            {
                Console.WriteLine("No tasks file found. Starting fresh.");
            }
        }

        private List<Task> LoadTasks()
        {
            if (!File.Exists(FileName)) return new List<Task>();
            var data = JsonConvert.DeserializeObject<Dictionary<string, List<Task>>>(File.ReadAllText(FileName));
            return data != null && data.ContainsKey("tasks") ? data["tasks"] : new List<Task>();
        }

        private void SaveTasks(List<Task> tasks)
        {
            var data = new Dictionary<string, List<Task>> { { "tasks", tasks } };
            File.WriteAllText(FileName, JsonConvert.SerializeObject(data, Formatting.Indented));
        }

        public void AddTask(string description, string mark)
        {
            var tasks = LoadTasks();
            tasks.Add(new Task { ID = _id++, Description = description, Mark = mark });
            SaveTasks(tasks);
            Console.WriteLine("Task added successfully.");
        }

        public void UpdateTask(int taskId, string newDescription)
        {
            var tasks = LoadTasks();
            var task = tasks.Find(t => t.ID == taskId);
            if (task != null)
            {
                task.Description = newDescription;
                SaveTasks(tasks);
                Console.WriteLine($"Task {taskId} updated successfully.");
            }
            else
            {
                Console.WriteLine($"Task {taskId} not found.");
            }
        }

        public void DeleteTask(int taskId)
        {
            var tasks = LoadTasks();
            var originalCount = tasks.Count;
            tasks.RemoveAll(t => t.ID == taskId);
            SaveTasks(tasks);

            if (tasks.Count < originalCount)
                Console.WriteLine($"Task {taskId} deleted successfully.");
            else
                Console.WriteLine($"Task {taskId} not found.");
        }

        public void MarkTask(int taskId, string mark)
        {
            var tasks = LoadTasks();
            var task = tasks.Find(t => t.ID == taskId);
            if (task != null)
            {
                task.Mark = mark;
                SaveTasks(tasks);
                Console.WriteLine($"Task {taskId} marked as {mark}.");
            }
            else
            {
                Console.WriteLine($"Task {taskId} not found.");
            }
        }

        public void ShowTasks(string filter)
        {
            var tasks = LoadTasks();
            var filteredTasks = filter == "all" ? tasks : tasks.FindAll(t => t.Mark == filter);

            foreach (var task in filteredTasks)
            {
                Console.WriteLine($"ID: {task.ID}, Description: {task.Description}, Mark: {task.Mark}");
            }
        }
    }
}
